keps/sig-cluster-lifecycle/kubeadm-config-ha-support.md

```diff
@@ -0,0 +1,149 @@
+     * [Risks and Mitigations](#risks-and-mitigations)
+  * [Implementation History](#implementation-history)
+  * [Drawbacks](#drawbacks)
+  * [Alternatives](#alternatives)
+
+## Summary
+
+Currently, `kubeadm upgrade` of a master in a multi-master cluster will fail without workarounds because of a lack of node-specific master configuration in the `kubeadm-config` ConfigMap that is created at `init` time and later referenced during an upgrade.
+In particular, node-specific information is required in order for kubeadm to identify which control plane static pods belong to the current node during an upgrade.
+In the non-HA case, having a single `nodeName` property in `kubeadm-config` that corresponds to the single master is sufficient because there is no ambiguity.
+As we move towards supporting HA natively in kubeadm, a new approach is required to uniquely identify master nodes.
+
+## Motivation
+
+Kubeadm is driving towards natively supporting highly available clusters.
+As part of HA support, a clean upgrade path is required.
+The purpose of this KEP is simply to introduce support for multiple masters in the kubeadm configuration that is stored in-cluster in order to enable that clean upgrade path.
+
+### Goals
+
+Enable `kubeadm upgrade` of highly available clusters by augmenting the existing persistent kubeadm configuration.
+
+### Non-Goals
+
+This proposal does not aim to solve the entire problem of upgrading HA clusters.
+This KEP specifically tackles the persistent configuration problem so that the information required at upgrade time is available.
+
+### Challenges and Open Questions
+
+The final implementation of this KEP will require deciding exactly what "master node-specific information" means.
+Currently, the `nodeName` of the master is the only entry that is unarguably node-specific.
+However, it may be possible that additional config entries could be split out into the node-specific area(s) of the config.
+This could result in asymmetric configuration across the masters, which may or may not be something that we wish to support.
+
+## Proposal
+
+### Implementation Details
+
+#### Background
+
+Currently, the `kubeadm-config` ConfigMap in the `kube-system` namespace serves as the single source of truth for how kubeadm has been used to create and modify a cluster.
+Because kubeadm is not a process that runs on the cluster (it is only run to perform operations, e.g. `init` and `upgrade`), this config is not modified during normal operation.
+In the non-HA case today, it is guaranteed to be an accurate representation of the kubeadm configuration.
+
+If kubeadm is used to create an HA cluster today, e.g. using the workarounds described in [kubeadm #546](https://github.com/kubernetes/kubeadm/issues/546) and/or @mbert's [document](https://docs.google.com/document/d/1rEMFuHo3rBJfFapKBInjCqm2d7xGkXzh0FpFO0cRuqg), then the `kubeadm-config` ConfigMap will be an accurate representation except for any master node-specific information.
```

Comment:
shouldn't we refer to the actual HA doc too?

Reply:
I decided that it would be better to leave this as-is since it would be difficult to shoehorn in a reference to the HA doc here and there are plenty of links out to it from the workarounds links. It also didn't seem to fit in well with the other link sections (`see-also` which is ideally just to other KEPs and `Implementation History`).

Reply:
why is it difficult to add a link?

Reply:
I added a link to the official kubeadm HA doc in the `see-also` section.